### ISSUE TITLE
Make map bias more dynamic 

### ISSUE THREAD

**Issue by rmarquis** [etlegacy/etlegacy#2317]
Using total amount of maps played and wins in map bias calculation will eventually end up in static number once total number goes big enough. 

Please refer to this post for full analysis of the issue and possible solution.
_Originally posted by @u1ps in https://github.com/etlegacy/etlegacy/discussions/2067_

As stated above, the first step to achieve is updating the database scheme to allow for the right data to be collected. As discussed, other fields useful for further improvement of the existing skill rating system (see #403) could be also added, as to adjust the database scheme only a single time.

The required fields are as follow, for the respective improvements discussed:

Dynamic map bias:
* adding win/loss numbers but as float, not integer
* the integer field could be dismissed (maybe keep it and use it to store the ever increasing, raw values)
* adding a way to skip "no objective" maps, without winner/loser/draw, when admin force the end of maps instead of waiting for one of the team to win (see #3066)

Time decay improvement:
* storing timestamp of last time the map was played
* changing last time a player has its rating updated from text to timestamp
* the text field might still be useful for quick debug purpose, so maybe keep it?
* also add map last play time as text for the same reason?

For adding the map vote system in the database (as opposed to the current separate, file based system):
* add number of time the map was played (integer)
* number of other maps played since it was last played

For monitoring of the system, add performance stats to check the team winning prediction is on a given map
* confusion matrix (TP, FP, TN, FN) as 4 integer fields, to derive precision/recall and F1-score for each map
* do the same for global stats (in a new table?)
* add some sort of time decay for these stats too? (last N maps played representing last M months). Not sure how to implement that efficiently though.

Once this is done, the second step to fix the map bias becoming gradually static to make it dynamic:

* keep the initial continuity correction of map = 20
* once all fake data is gone, let map win/loss numbers grow to 200 plays (N = 200 seems to be an appropriate approximative value, representing how much time of data is kept for 1 month.)
* downscale to 200 and ignore older data to compute the map bias.

Note: store the map win ratio in the database normalized to 100 (and then multiplied by N = 200 in the code), so in case we want to change it later in the code, the stored data can be reused without any issue.

As further improvement (but later on), N could be computed totally dynamically with time decay. The same could be done with players rating decay with the Tau parameter. See Glicko 2 that increase uncertainty due to inactivity.

**Comment by rmarquis** [comment]
Some more resources/discussion about time rating decay in TrueSkill: See https://github.com/bernd-wechner/CoGs/issues/13

Notably, decay could be modelled as a CDF of a Laplace distribution, for maps rating, but also player rating. The idea is to implement a decline in confidence that would ultimetly far better model expectation than the actual Tau linear uncertainty (set to default value in ETL).

In practice, this could potentially greatly mitigate top players with high confidence "farming" new players and hardly going lower in the skill ranking.

**Comment by rmarquis** [comment]
The following comment documents the implementation plan for the dynamic map bias, with each constant derived statistically rather than chosen arbitrarily, in the same spirit as the original [Bayesian skill rating design doc](https://etlegacy.readthedocs.io/en/latest/skillrating.html).

(This level of detail is mostly for my own sanity. I do live in my own world, and you're all just visiting.)

## Proposal: dynamic map bias via capped EMA (N = 200) - narrowed scope

Following the above, we'll implement the dynamic map bias as a focused change, deferring the other schema additions from the issue body (map vote tracking, prediction confusion matrix, player rating timestamps) to separate issues - one schema change at a time keeps each migration reviewable and avoids feature creep.

### Algorithm: exponential moving average over plays, with exact cap

The rating_maps win counters become floats and act as a decayed evidence sum. On each map end with a decisive winner:

1. If win_axis_f + win_allies_f >= N, scale both counters by (N − 1) / total - making room for exactly one new result.
2. Increment the winner's counter by 1.

The post-update total is therefore exactly N, and each past result's weight decays by a factor of (N−1)/N per play - a smooth exponential forgetting over ~N plays, instead of today's unbounded counters where a new result moves the bias by 1/total (→ 0 as the database ages, the freeze described in this issue).

### Justification of N = 200

N is bounded from two sides. With decayed counters summing to N, the bias estimate has standard error sqrt(p(1−p)/N), worst case 1/(2√N) at p = 0.5 - too small an N and the computed "bias" is mostly sampling noise, jittering the map term in every rating update. In the other direction, the EMA tracks a genuine change in map outcomes with a lag of ~N plays - too large an N and the bias again reflects an old reality. N = 200 sits at the statistical lower bound: maximal responsiveness subject to the estimate not being noise.

The required accuracy is derived from the rating system's own noise floor, not asserted: map bias shifts expected team performance by 2·MU·ε = 50ε mu, and player sigma is MU/3 ≈ 8.3. Estimate errors below half a sigma (ε ≤ ~8%) cannot meaningfully corrupt ratings - the model already carries that much uncertainty. This bounds the required accuracy to roughly ±5–8%; δ = 7% is the conservative end (8% would give N = 150, 4% would push past N = 600 polishing a term the model can't feel). Cross-check: known stock-map skews run 55–70%, well above the 7% detection threshold, so every bias that matters is caught.

The 95% confidence level (z = 1.96) is conventional, but the result is insensitive to it: 90% → N ≈ 138, 95% → N ≈ 196, 99% → N ≈ 338. N = 200 is a representative of a robust range, not a brittle point estimate.

```
  N = (1.96 / (2 × 0.07))² ≈ 196 → rounded to 200
```

The iid-binomial error model is itself conservative: the EMA's geometric weights (steady state 1, r, r², … with r = 199/200) give an effective sample size of (Σwᵢ)²/Σwᵢ² ≈ 2N ≈ 400 against a static bias, so true accuracy is ~2.5% rather than 3.5% - the approximation costs us nothing and keeps the derivation simple.

This is also consistent with the existing continuity correction: LAMBDA = 10 fake wins per side is a Beta(10,10) prior = 20 fake matches, bridging exactly to the n > 20 regime where the normal approximation is valid (per the skillrating docs). The two mechanisms hand off cleanly - LAMBDA dominates while data is scarce, the cap keeps responsiveness constant once data is plentiful. A ±7% bias error translates to ~3.5 mu (≈ 0.42 sigma) of rating impact, well under the rating system's own noise floor.

### Database changes (schema version 3)

```sql
  ALTER TABLE rating_maps ADD COLUMN win_axis_f REAL;
  ALTER TABLE rating_maps ADD COLUMN win_allies_f REAL;
  ALTER TABLE rating_maps ADD COLUMN last_played INTEGER;  -- unix time
```

* win_axis / win_allies (INT) are kept and still incremented as raw lifetime totals (audit/debug, migration source).
* win_axis_f / win_allies_f are backfilled from the raw counters at migration, scaled down to N if raw totals exceed it.
* last_played is populated from day one but not read by any code path yet - it exists so the follow-up time-decay variant is a code-only change with real data waiting, no second migration.

### Bundled small fixes (same feature scope)

* Draws / aborted maps (no decisive winner) are no longer recorded as Axis wins - see #3066.
* G_SkillRatingGetMapRating returns 0.5 (fair) instead of 1.0 on DB errors.

### Follow-up (not part of this change): time decay

The capped EMA's memory is measured in plays, not calendar time: a map idle for 8 months returns with a fully confident bias computed from a possibly outdated player population. Time decay (Glicko-2-style uncertainty growth, or the Laplace-CDF model discussed above) would shrink the effective sample size during idle periods, re-engaging the prior toward 0.5 - the system becomes humble about stale data.

#### Design notes:

* It replaces the cap's scaling factor with exp(−Δt/τ) (using last_played); all float/correction machinery is reused.
* τ would be calibrated so the effective sample size (Σwᵢ)²/Σwᵢ² ≈ 200 at typical play rates, preserving the ±7% detectability target.
* To be implemented only if stale-bias problems are observed on low-rotation servers; a cheap middle ground is a staleness reset (re-apply continuity correction when now − last_played exceeds a threshold).

## Relation to the original proposal above

This plan implements the core of step 2 as originally proposed (float win/loss counters, growth capped at N ≈ 200 with downscaling of older data, initial continuity correction kept), with two deliberate deviations:

1. Schema migrations are split per feature instead of bundled. The original proposal suggested adjusting the database scheme a single time to collect fields for dynamic map bias, time decay, map vote tracking, and prediction monitoring together. We'll instead pay the migration cost once per feature: each change ships isolated, reviewable, and revertible, and a problem in one feature can't stall the others. The map bias migration (schema v3) does include the last_played timestamp column - populated from day one, read by no code path yet - specifically so the later time-decay step remains a code-only change with real data already collected, i.e. the bundling concern is addressed where it's cheap without pulling unrelated features in.

2. Decayed counts are stored directly instead of a ratio normalized to 100. The original proposal suggested storing the win ratio normalized to 100 and multiplying by N in code, so N could be retuned without invalidating stored data. We store the decayed win/loss counts as floats summing to N.

    Trade-off: retuning N later requires rescaling stored counts (a one-time SQL update or on-the-fly rescale at read) rather than changing a single constant. We judge this acceptable because storage and computation share one unit (no conversion bugs), the data stays directly inspectable (win_axis_f + win_allies_f = 200 is self-documenting in a DB dump), and N's derivation (±7% detectability at 95% confidence) is statistical rather than load-dependent, so retuning is unlikely.

    The statistical framing also refines the original "N = 200 ≈ one month of data" estimate: N = (1.96 / (2 × 0.07))² ≈ 196, i.e. the smallest effective sample that reliably detects a ±7% map bias at 95% confidence. The calendar-time interpretation is kept as intuition only, since plays-per-month varies per server.

Everything else follows the issue as written: raw INT counters kept as lifetime totals, time decay (Glicko-2-style or Laplace-CDF) deferred as a follow-up, and the no-objective/draw guard from #3066 included since it directly affects what enters the bias counters.

**Comment by u1ps** [comment]
Does SR system take into account player count (estimate % chance)? Currently there seems to be a trendline of win chance going lower for attacking team when player amount in server increases. 

Global win rate for Allies in TM server is 51.5% but when take data of matches with 36 players or more then the number becomes 40%.

If there is not player amount dependency then it means currently in TM server when playing during peak time in defensive team you have better chance of increasing SR.

**Comment by rmarquis** [comment]
The win probability is purely a function of the sum of skills (represented by an estimate value, and an uncertainty of that value) of players of each team, and the estimated map bias. That sum (team performance) is assumed to be additive and linear, which is certainly not the case in reality, but this is a simplified model.

So to answer your question directly: assuming the teams are balanced in player count (18 vs 18 players), no, the 40% vs 51.5% you observed is not directly related to the number of players but it is about the quality of their summed (estimated) skill. Keep in mind the map bias is also less of a factor with more players though, but that map bias estimation is also dependent on the number of players.

TrueSkill is very well documented if you are interested. The bottom line is, however, that there is no easy explanation of it unless you get into the maths behind the algorithm. Each factor is interdependent in a self-balancing and self-correcting way.
